src/ast/nodes/Program.ts

```diff
@@ -2,11 +2,29 @@ import MagicString from 'magic-string';
 import { NodeBase, StatementNode } from './shared/Node';
 import { NodeType } from './NodeType';
 import { RenderOptions, renderStatementList } from '../../utils/renderHelpers';
+import ExecutionPathOptions from '../ExecutionPathOptions';
 
 export default class Program extends NodeBase {
 	type: NodeType.Program;
 	body: StatementNode[];
 
+	hasEffects(options: ExecutionPathOptions) {
+		for (const node of this.body) {
+			if (node.hasEffects(options)) return true;
+		}
+	}
+
+	include() {
+		let anotherPassNeeded = false;
+		this.included = true;
+		for (const node of this.body) {
+			if (node.shouldBeIncluded() && node.include()) {
+				anotherPassNeeded = true;
```

Comment:
minor opt: does this require a `break` when it finds at least one `true` clause?

Reply:
No, if we break once the first `include` states that we need another pass, this will destroy performance. `include` is meant to include everything we know we need based on the currently included variables. The return value is basically a flag indicating that new variables have been added and previous includes could possibly include more statements.

However if you break as soon as a variable is added, a new tree-shaking pass will start *for each added variable*, no matter how local it is. This will result in hundreds of tree-shaking passes instead of maybe a dozen. Even though those passes will on average be twice as fast (due to the early bail out), this will basically stall everything. The current goal is to instead include as much as possible as early as possible to reduce the number of passes. Also for many types of included statements, the side-effect detection can quickly skip over these statements.

However I agree that it is confusing that a method that has side-effects (i.e. including the children) also has a return value. I think I should rework this to instead let the nodes that add variables set a "dirty" flag on the module to indicate another pass is needed.